File: tugas_Heap/src/MinHeapData.java
import java.util.ArrayList;

/**
 * MinHeapData class - Implementasi Min-Heap untuk DataEntry (sorted ascending by id)
 * Menyediakan operasi: insert, delete, peek, dan display
 */
public class MinHeapData {
    private ArrayList<DataEntry> heap;

    /**
     * Constructor - inisialisasi heap kosong
     */
    public MinHeapData() {
        this.heap = new ArrayList<>();
    }

    /**
     * Operasi INSERT - menambahkan elemen ke min-heap
     * Kompleksitas: O(log n)
     */
    public void insert(DataEntry entry) {
        heap.add(entry);
        int index = heap.size() - 1;

        // Heapify-up: bandingkan dengan parent dan swap jika perlu
        while (index > 0) {
            int parentIndex = (index - 1) / 2;
            if (heap.get(index).compareTo(heap.get(parentIndex)) < 0) {
                swap(index, parentIndex);
                index = parentIndex;
            } else {
                break;
            }
        }
    }

    /**
     * Operasi PEEK - mendapatkan elemen terkecil (root) tanpa menghapusnya
     * Kompleksitas: O(1)
     */
    public DataEntry peek() {
        if (heap.isEmpty()) {
            return null;
        }
        return heap.get(0);
    }

    /**
     * Operasi DELETE MIN - menghapus elemen terkecil (root)
     * Kompleksitas: O(log n)
     */
    public DataEntry deleteMin() {
        if (heap.isEmpty()) {
            return null;
        }

        DataEntry min = heap.get(0);
        heap.set(0, heap.get(heap.size() - 1));
        heap.remove(heap.size() - 1);

        if (!heap.isEmpty()) {
            heapifyDown(0);
        }

        return min;
    }

    /**
     * Operasi DELETE - menghapus entry spesifik dari heap
     * Kompleksitas: O(n)
     */
    public boolean delete(int id) {
        int index = findIndex(id);

        if (index == -1) {
            return false;
        }

        heap.set(index, heap.get(heap.size() - 1));
        heap.remove(heap.size() - 1);

        if (index < heap.size()) {
            heapifyDown(index);
        }

        return true;
    }

    /**
     * Helper: Heapify-down - mempertahankan sifat min-heap ke bawah
     * Kompleksitas: O(log n)
     */
    private void heapifyDown(int index) {
        int smallest = index;
        int leftChild = 2 * index + 1;
        int rightChild = 2 * index + 2;

        if (leftChild < heap.size() && heap.get(leftChild).compareTo(heap.get(smallest)) < 0) {
            smallest = leftChild;
        }

        if (rightChild < heap.size() && heap.get(rightChild).compareTo(heap.get(smallest)) < 0) {
            smallest = rightChild;
        }

        if (smallest != index) {
            swap(index, smallest);
            heapifyDown(smallest);
        }
    }

    /**
     * Helper: Tukar dua elemen di heap
     */
    private void swap(int i, int j) {
        DataEntry temp = heap.get(i);
        heap.set(i, heap.get(j));
        heap.set(j, temp);
    }

    /**
     * Helper: Cari index entry dengan id tertentu
     */
    private int findIndex(int id) {
        for (int i = 0; i < heap.size(); i++) {
            if (heap.get(i).getId() == id) {
                return i;
            }
        }
        return -1;
    }

    /**
     * Utility: Cek apakah heap kosong
     */
    public boolean isEmpty() {
        return heap.isEmpty();
    }

    /**
     * Utility: Dapatkan ukuran heap
     */
    public int size() {
        return heap.size();
    }

    /**
     * Utility: Dapatkan semua data dalam urutan min-heap (ascending)
     */
    public ArrayList<DataEntry> getAllSorted() {
        ArrayList<DataEntry> result = new ArrayList<>();
        MinHeapData tempHeap = new MinHeapData();

        // Copy semua elemen
        for (DataEntry entry : heap) {
            tempHeap.insert(entry);
        }

        // Extract semua elemen dalam urutan ascending
        while (!tempHeap.isEmpty()) {
            result.add(tempHeap.deleteMin());
        }

        return result;
    }

    /**
     * Utility: Display heap
     */
    public void display() {
        if (heap.isEmpty()) {
            System.out.println("Min-Heap kosong!");
            return;
        }

        ArrayList<DataEntry> sorted = getAllSorted();
        System.out.println("┌─────┬─────────────────────────────┐");
        System.out.println("│ ID  │ NAMA                        │");
        System.out.println("├─────┼─────────────────────────────┤");
        for (DataEntry entry : sorted) {
            System.out.println("│ " + entry.toString() + " │");
        }
        System.out.println("└─────┴─────────────────────────────┘");
    }

    /**
     * Utility: Get heap list (internal structure)
     */
    public ArrayList<DataEntry> getHeapList() {
        return new ArrayList<>(heap);
    }
}

File: tugas_Heap/heap_py/min_heap_data.py
from data_entry import DataEntry


class MinHeapData:
    """Implementasi Min-Heap untuk DataEntry, diurutkan ascending berdasarkan id."""

    def __init__(self):
        self.heap: list[DataEntry] = []

    def insert(self, entry: DataEntry) -> None:
        """Menambahkan elemen ke min-heap. Kompleksitas: O(log n)."""
        self.heap.append(entry)
        self._heapify_up(len(self.heap) - 1)

    def peek(self) -> DataEntry | None:
        """Mengambil elemen terkecil tanpa menghapusnya. Kompleksitas: O(1)."""
        if self.is_empty():
            return None
        return self.heap[0]

    def delete_min(self) -> DataEntry | None:
        """Menghapus elemen terkecil/root. Kompleksitas: O(log n)."""
        if self.is_empty():
            return None

        min_entry = self.heap[0]
        last_entry = self.heap.pop()

        if not self.is_empty():
            self.heap[0] = last_entry
            self._heapify_down(0)

        return min_entry

    def delete(self, id_data: int) -> bool:
        """Menghapus entry spesifik berdasarkan id. Kompleksitas: O(n)."""
        index = self._find_index(id_data)

        if index == -1:
            return False

        last_entry = self.heap.pop()

        if index < len(self.heap):
            self.heap[index] = last_entry
            self._fix_heap_after_replace(index)

        return True

    def is_empty(self) -> bool:
        return len(self.heap) == 0

    def size(self) -> int:
        return len(self.heap)

    def get_all_sorted(self) -> list[DataEntry]:
        """Mengambil semua data dalam urutan ascending tanpa merusak heap asli."""
        result: list[DataEntry] = []
        temp_heap = MinHeapData()

        for entry in self.heap:
            temp_heap.insert(entry)

        while not temp_heap.is_empty():
            result.append(temp_heap.delete_min())

        return result

    def display(self) -> None:
        if self.is_empty():
            print("Min-Heap kosong!")
            return

        sorted_entries = self.get_all_sorted()
        print("+-------+--------------------------------+")
        print("| ID    | NAMA                           |")
        print("+-------+--------------------------------+")
        for entry in sorted_entries:
            print(f"| {entry} |")
        print("+-------+--------------------------------+")

    def get_heap_list(self) -> list[DataEntry]:
        return self.heap.copy()

    def _heapify_up(self, index: int) -> None:
        while index > 0:
            parent_index = (index - 1) // 2
            if self.heap[index].compare_to(self.heap[parent_index]) < 0:
                self._swap(index, parent_index)
                index = parent_index
            else:
                break

    def _heapify_down(self, index: int) -> None:
        while True:
            smallest = index
            left_child = 2 * index + 1
            right_child = 2 * index + 2

            if (
                left_child < len(self.heap)
                and self.heap[left_child].compare_to(self.heap[smallest]) < 0
            ):
                smallest = left_child

            if (
                right_child < len(self.heap)
                and self.heap[right_child].compare_to(self.heap[smallest]) < 0
            ):
                smallest = right_child

            if smallest == index:
                break

            self._swap(index, smallest)
            index = smallest

    def _fix_heap_after_replace(self, index: int) -> None:
        parent_index = (index - 1) // 2
        if index > 0 and self.heap[index].compare_to(self.heap[parent_index]) < 0:
            self._heapify_up(index)
        else:
            self._heapify_down(index)

    def _swap(self, i: int, j: int) -> None:
        self.heap[i], self.heap[j] = self.heap[j], self.heap[i]

    def _find_index(self, id_data: int) -> int:
        for index, entry in enumerate(self.heap):
            if entry.get_id() == id_data:
                return index
        return -1
